app_build_suite/__main__.py

```diff
@@ -1,8 +1,138 @@
+        help=f"List of steps to execute. Available steps: {ALL_STEPS}",
+        required=False,
+        default=["all"],
+    )
+
+
+def get_global_config_parser(add_help: bool = True) -> configargparse.ArgParser:
+    config_parser = configargparse.ArgParser(
+        prog=app_name,
+        add_config_file_help=True,
+        default_config_files=[".abs.yaml"],
+        description="Build and test Giant Swarm App Platform app.",
+        add_env_var_help=True,
+        auto_env_var_prefix="ABS_",
+        formatter_class=configargparse.ArgumentDefaultsHelpFormatter,
+        add_help=add_help,
+    )
+    configure_global_options(config_parser)
+    return config_parser
+
+
+def validate_global_config(config: configargparse.Namespace):
+    # validate build engine
+    if config.build_engine not in ALL_BUILD_ENGINES:
+        raise ConfigError(
+            "build_engine",
+            f"Unknown build engine '{config.build_engine}'. Valid engines are: {ALL_BUILD_ENGINES}.",
+        )
+    # validate steps
+    for step in config.steps:
+        if step not in ALL_STEPS:
+            raise ConfigError(
+                "steps", f"Unknown step '{step}'. Valid steps are: {ALL_STEPS}."
+            )
+
+
+def get_config(steps: List[BuildStep]) -> configargparse.Namespace:
+    # initialize config, setup arg parsers
+    try:
+        config_parser = get_global_config_parser()
+        for step in steps:
+            step.initialize_config(config_parser)
+        config = config_parser.parse_args()
+        validate_global_config(config)
+    except ConfigError as e:
+        logger.error(f"Error when checking config option '{e.config_option}': {e.msg}")
+        sys.exit(1)
+
+    logger.info("Starting build with the following options")
+    logger.info(f"\n{config_parser.format_values()}")
+    return config
+
+
+def run_pre_steps(config: configargparse.Namespace, steps: List[BuildStep]) -> None:
+    for step in steps:
+        step.pre_run(config)
+
+
+def run_build_steps(config: configargparse.Namespace, steps: List[BuildStep]) -> None:
+    for step in steps:
+        step.run(config)
+
+
+def run_cleanup(config: configargparse.Namespace, steps: List[BuildStep]) -> None:
+    for step in steps:
+        step.cleanup(config)
 
 
 def main():
-    pass
+    log_format = "%(asctime)s %(name)s %(levelname)s: %(message)s"
+    logging.basicConfig(format=log_format)
+    logging.getLogger().setLevel(logging.INFO)
+
+    global_only_config_parser = get_global_config_parser(add_help=False)
+    global_only_config = global_only_config_parser.parse_known_args()[0]
+    if global_only_config.debug:
+        logging.getLogger().setLevel(logging.DEBUG)
+
+    container = ComponentsContainer()
+    container.config.from_dict({"build_engine": global_only_config.build_engine},)
+
+    steps = get_pipeline(container)
+    config = get_config(steps)
+    run_pre_steps(config, steps)
+    run_build_steps(config, steps)
+    run_cleanup(config, steps)
```

Comment:
Would it make sense to always execute `run_cleanup` even if something failed before?

Reply:
Good idea, probably with a flag informing if the build was OK or fail (you might want to leave something for debugging if the build was failed).